src/python/gem5/simulate/multi_sim.py

```diff
@@ -0,0 +1,113 @@
+# notice, this list of conditions and the following disclaimer;
+# redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution;
+# neither the name of the copyright holders nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import itertools
+import os
+import signal
+from functools import partial
+from time import sleep
+from typing import (
+    Any,
+    Callable,
+    List,
+    Tuple,
+)
+
+from gem5.resources.resource import AbstractResource
+from gem5.utils.multiprocessing import Process
+
+from .simulator import Simulator
+
+
+class MultiSim:
+    """
+    This class is used to run multiple simulations in parallel. It takes a list of
+    callables that return a Simulator object and runs them in parallel. The number of
+    simulations that can be run in parallel is limited to the value of the num_cpus parameter
+    """
+
+    def __init__(self, sim_callables: List[Callable[[], Simulator]]):
+        """
+        :param sim_callables: A list of callables that return a Simulator object
+        """
+        self.sim_callables = sim_callables
+
+    def run_all(self, num_cpus: int):
+        """
+        Run all the simulations in parallel
+        :param num_cpus: The number of simulations to run in parallel at a time
+        """
+        processes = []
+
+        for sim_callable in self.sim_callables:
+            if len(processes) > num_cpus:
+                for process in processes:
+                    if not process.is_alive():
+                        processes.remove(process)
+                    os.kill(process.pid, signal.SIGTERM)
+                sleep(1)
+            process_name = (
+                f"{sim_callable.func.__name__}_{sim_callable.args[0].get_id()}"
+            )
+            process = Process(
+                target=run_simulator,
+                args=(sim_callable,),
+                name=process_name,
+            )
+            print(f"Starting process {process_name}")
```

Comment:
We shouldn't have any `print` statements in this code. There's a good argument to need *logging* but for that, we should use something better. E.g., `info()` if we can or `logging` in python. That said, we need to tread lightly here with changes. My only suggestion is to remove the print statement or change it to an `info`